src/components/structures/TimelinePanel.js

```diff
@@ -291,6 +299,21 @@ var TimelinePanel = React.createClass({
         this.forceUpdate();
     },
 
+    onLocalEchoUpdated: function(ev, room, oldEventId) {
+        if (this.unmounted) return;
+
+        // ignore events for other rooms
+        if (room !== this.props.room) return;
+
+        // Once the remote echo for an event arrives, we need to turn the
+        // greyed-out event black.  When the localEchoUpdated event is raised,
+        // the nested 'event' property within one of the events in
+        // _timelineWindow will have been replaced with the new event. So
+        // all that is left to do here is to make the message-panel re-render.
+        this.forceUpdate();
```

Comment:
Well, there's certainly nothing for the timelinepanel to do here.

The answer to your second question is basically just about the layering between the TimelinePanel and the MessagePanel.  I've deliberately kept all of the event-listening out of MessagePanel, to keep it simple. It can't listen for the event because it doesn't know what room it is showing.

We *could* do `this.refs.messagePanel.forceUpdate()` instead of `this.forceUpdate()`, but that's not very react-idiomatic.